src/mpid/ch4/src/ch4_impl.h

```diff
@@ -320,6 +320,19 @@ MPL_STATIC_INLINE_PREFIX void MPIDIG_win_hash_clear(MPIR_Win * win)
         }                                                       \
     } while (0)
 
+#define MPIDI_Datatype_get_density(datatype_, density_)         \
+    do {                                                        \
+        if (IS_BUILTIN(datatype_)) {                            \
+            (density_) = INT_MAX;                               \
+        } else {                                                \
+            MPIR_Datatype *dt_ptr_;                             \
+            MPIR_Datatype_get_ptr((datatype_), dt_ptr_);        \
+            MPI_Aint size_;                                     \
+            MPIR_Datatype_get_size_macro((datatype_), size_); \
+            (density_) = size_ / (dt_ptr_)->max_contig_blocks;	\
```

Comment:
What is `max_contig_blocks`? Is it the same as `MPIR_Typerep_get_iov_len`?

Reply:
I think we should put this macro in `mpir_datatype.h` now, because the context/concept for this code is independent of ofi device or any device. It's a datatype macro and I think it is easier to be maintained there.

On another note, if this information is useful (yes since we needed at runtime) then may be we should consider put this information in the datatype structure and compute it during Type_commit time. Then since it is out-of time critical places, we can afford for more accurate stat if necessary.

Reply:
I think `max_contig_blocks` only pertains to a single instance of the datatype. @pavanbalaji is that correct? I have no problem moving to `mpir_datatype.h`.

Reply:
Yes.  For example, if multiple datatypes are all contiguous, the total contig count should still be 1.  Similarly, if the end of one noncontig datatype and the start of the next are contiguous, those can be combined too.  Dataloop does those optimizations, though yaksa doesn’t.

FYI, your check for INT_MAX should be based on is_contig, not on BUILTIN datatypes.  Otherwise you’ll miss other contiguous types, such as dup of INT.

Reply:
@pavanbalaji Does `max_contig_blocks` return the same as `MPIR_Typerep_get_iov_len`?

Reply:
Typerep takes a count argument, and max_contig_block is per datatype.  So they are not equivalent.  But max_contig_block * count >= what Typerep returns.

Reply:
Thanks, it means I shouldn't use `MPIR_Typerep_get_iov_len` to calculate density.

Reply:
Actually, the opposite.  Typerep is the only accurate way to calculate the density, but it is more expensive to use than max_contig_block.  In this case, because an approximation is sufficient, max_contig_block is probably fine.

Reply:
To be clear: it’s more expensive for dataloop, but not for yaksa.

Reply:
@pavanbalaji what is your suggestion for packing threshold? This PR uses 128 bytes per segment. Shawn mentioned using something like 16K(?) in the GPU work, but that probably assumes higher data movement cost.

Reply:
@raffenet `ch3` uses 16K, which is why @shawnccx picked that value.  Of course, it needs to be tuned, but 16K might be good starting point for now.

Reply:
I don't think the default threshold matters for now. Either 128 or 16K is away from the extreme cases that we are trying to fix, so both or any should work as default. If we really care about the default performance, we need run measurements/study and then judge from the data.

Reply:
This PR should either run tests and find a good threshold or pick one that we have been using for `ch3`.  A different arbitrary number doesn't make sense.

Reply:
Got it. I will use `MPIR_CVAR_CH4_IOV_DENSITY_MIN`, which is based on `ch3` from what I can tell.

Reply:
Sounds good for now.  In the future, we might have different thresholds, at least for CPUs vs. GPUs, but this is fine for now.